package-build.el

```diff
@@ -192,262 +192,187 @@ or nil if the version cannot be parsed."
-                 argv exit-code (buffer-string))))))
-
-(defun package-build--run-process-match (regexp dir prog &rest args)
-  "Run PROG with args and return the first match for REGEXP in its output.
-PROG is run in DIR, or if that is nil in `default-directory'."
+(defun package-build--run-process (directory destination command &rest args)
+  (with-current-buffer
+      (if (eq destination t)
+          (current-buffer)
+        (or destination (get-buffer-create "*package-build-checkout*")))
+    (let ((default-directory
+            (file-name-as-directory (or directory default-directory)))
+          (argv (nconc (unless (eq system-type 'windows-nt)
+                         (list "env" "LC_ALL=C"))
+                       (if package-build-timeout-executable
+                           (nconc (list package-build-timeout-executable
+                                        "-k" "60" (number-to-string
+                                                   package-build-timeout-secs)
+                                        command)
+                                  args)
+                         (cons command args)))))
+      (unless (file-directory-p default-directory)
+        (error "Can't run process in non-existent directory: %s" default-directory))
+      (let ((exit-code (apply 'process-file
+                              (car argv) nil (current-buffer) t
+                              (cdr argv))))
+        (or (zerop exit-code)
+            (error "Command '%s' exited with non-zero status %d: %s"
+                   argv exit-code (buffer-string)))))))
+
+(defun package-build--run-process-match (regexp directory command &rest args)
   (with-temp-buffer
-    (apply 'package-build--run-process dir prog args)
+    (apply 'package-build--run-process directory t command args)
     (goto-char (point-min))
     (re-search-forward regexp)
     (match-string-no-properties 1)))
 
+(defun package-build--process-lines (directory command &rest args)
+  (with-temp-buffer
+    (apply 'package-build--run-process directory t command args)
+    (split-string (buffer-string) "\n" t)))
+
 ;;; Checkout
 ;;;; Common
 
-(defun package-build-checkout (name config working-dir)
-  "Check out source for the package named NAME with CONFIG under WORKING-DIR.
-In turn, this function uses the :fetcher option in the CONFIG to
-choose a source-specific fetcher function, which it calls with
-the same arguments.
-
-Returns the package version as a string."
-  (let ((fetcher (plist-get config :fetcher)))
-    (package-build--message "Fetcher: %s" fetcher)
-    (package-build--message "Source: %s\n"
-                            (or (plist-get config :repo)
-                                (plist-get config :url)))
-    (funcall (intern (format "package-build--checkout-%s" fetcher))
-             name config (file-name-as-directory working-dir))))
+(cl-defmethod package-build--checkout :before ((rcp package-recipe))
```

Comment:
Yes I know but ... ah I see you are talking about the `:before`. This code could also be moved into `package-build-archive`. I would prefer to leave it like this for a while though. It just messages some information and is rather irrelevant to the flow of things, which makes it desirable to separate it from the important parts. Once those parts are finished, the messages can be moved to the place that eventually turns out to be the correct one.

Reply:
Actually I think I would have liked to remove these messages completely but kept them to avoid unnecessary changes.

Reply:
This is the context in which these three lines are printed:

```
 • Building package magit ...
Package: magit
Fetcher: github
Source:  https://github.com/magit/magit.git

Cloning https://github.com/magit/magit.git to /home/jonas/git/emacs/melpa/working/magit/
```

* `Package: magit` is unnecessary, the same information is printed right above.
* `Source:  https://github.com/magit/magit.git` is unnecessary, the same information is printed right below.
* `Fetcher: github`, while not redundant, also isn't really useful here. This is not the place were one would make sure that a recipe author used e.g. `github` instead of the generic `git`. Also despite this output you (very seldom, but still) overlook such a mistake. I do notice and fix them later on, but I rely on other tools to do so.

So let's just remove this.